bundler/lib/bundler/fetcher.rb

```diff
@@ -148,14 +148,11 @@ def specs(gem_names, source)
       specs.each do |name, version, platform, dependencies, metadata|
         spec = if dependencies
           EndpointSpecification.new(name, version, platform, self, dependencies, metadata).tap do |es|
-            unless index.local_search(es).empty?
-              # Duplicate spec.full_names, different spec.original_names
-              # index#<< ensures that the last one added wins, so if we're overriding
-              # here, make sure to also override the checksum, otherwise downloading the
-              # specs (even if that version is completely unused) will cause a SecurityError
-              source.checksum_store.delete_full_name(es.full_name)
-            end
-            source.checksum_store.register(es, [es.checksum]) if source && es.checksum
+            # Duplicate spec.full_names, different spec.original_names
+            # index#<< ensures that the last one added wins, so if we're overriding
+            # here, make sure to also override the checksum, otherwise downloading the
+            # specs (even if that version is completely unused) will cause a SecurityError
+            source.checksum_store[es.full_name] = es.checksum if source && es.checksum
```

Comment:
Maybe it would work if we didn't guard on es.checksum here and allowed nil checksum to delete without setting.